pkg/cache/cache.go

```diff
@@ -76,22 +76,27 @@ func init() {
 // Get will return the cached configuration of the given ServiceAccount.
 // It will first look at the set of ServiceAccounts configured using annotations. If none are found, it will look for any
 // ServiceAccount configured through the pod-identity-webhook ConfigMap.
-func (c *serviceAccountCache) Get(name, namespace string) (role, aud string, useRegionalSTS bool, tokenExpiration int64) {
+func (c *serviceAccountCache) Get(name, namespace string) (role, aud string, useRegionalSTS bool, tokenExpiration int64, err error) {
```

Comment:
Would it be cleaner and easy to read if you just do something like?

```
func (c *serviceAccountCache) Get(name, namespace string) (role, aud string, useRegionalSTS bool, tokenExpiration int64, err error) {
	klog.V(5).Infof("Fetching sa %s/%s from cache", namespace, name)
	var respSA *CacheResponse
	{
		respSA = c.getSA(name, namespace)
		klog.V(5).Infof("Service account %s/%s not found in cache", namespace, name)
		if respSA != nil && respSA.RoleARN != "" {
			return respSA.RoleARN, respSA.Audience, respSA.UseRegionalSTS, respSA.TokenExpiration, nil
		} else if respSA != nil {
                        return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)
                }
	}
	{
		resp := c.getCM(name, namespace)
		if resp != nil {
			return resp.RoleARN, resp.Audience, resp.UseRegionalSTS, resp.TokenExpiration, nil
		} else {
                         return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)
                 }
	}
	//if no mutation is needed due to annotations return nil err to allow request through
	if respSA != nil && respSA.RoleARN == "" {
		return "", "", false, pkg.DefaultTokenExpiration, nil
	}
	return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)
}
```

Reply:
The problem with this is 
```if respSA != nil && respSA.RoleARN != "" {
			return respSA.RoleARN, respSA.Audience, respSA.UseRegionalSTS, respSA.TokenExpiration, nil
		} else if respSA != nil {
                        return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)
                }
                ```
                is their previous logic is relying on the precedent of 1. check SA 2. check CM 3. then respond so by returing before checking CM it breaks some logic

Reply:
Yeah I had something similar to that but in their logic they are giving equal precedence to SA as CM
they're expecting 
1. Check SA
2. Check CM
3. Allow/Deny Req
so if we short circuit the return in SA block but it is actually in CM block we miss that

Reply:
Oh I see, yeah this makes sense.

Reply:
I think this newest commit makes it cleaner 
```
	var respSA *CacheResponse
 	{
 		respSA = c.getSA(name, namespace)
 		if respSA != nil && respSA.RoleARN != "" {
 			return respSA.RoleARN, respSA.Audience, respSA.UseRegionalSTS, respSA.TokenExpiration, nil
 		}
 	}
 	{
 		respCM := c.getCM(name, namespace)
 		if respCM != nil {
 			return respCM.RoleARN, respCM.Audience, respCM.UseRegionalSTS, respCM.TokenExpiration, nil
 		}
 	}
 	if respSA != nil && respSA.RoleARN == "" {
 		return "", "", false, pkg.DefaultTokenExpiration, nil
 	}
 	return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)

 }

```

Reply:
oh I didn't see your edit @StrongMonkey what do you think of this way

Reply:
great! I would suggest this in the end of the three line to make it more clear.

```
if respSA == nil {
   return "", "", false, pkg.DefaultTokenExpiration, fmt.Errorf("service account %s/%s not found in cache and one is expected", namespace, name)
}

return "", "", false, pkg.DefaultTokenExpiration, nil
```

This is so that when people read this it is clear that we will return error when respSa is nil, which means that it is missing from cache.

Reply:
agreed should be updated now